### PR TITLE
[GOBBLIN-1212] Handle non-Primitive type eligibility-check for shuffle key properly

### PR DESCRIPTION
**Pull request by autumnust**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [ ] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1212

### Description
- [ ] Here are some details about my PR, including screenshots (if applicable):

Before this change, the handling for non-primitive types like map,array and union will hit NPE since they don't have `fieldNames`. This fixed that and making sure the eligibility check happens recursively into those types.

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-compaction/src/main/java/org/apache/gobblin/compaction/mapreduce/orc/OrcUtils.java [hunk 1]
```diff
@@ -491,18 +491,34 @@ public static WritableComparable createValueRecursively(TypeDescription schema)
    */
   public static boolean eligibleForUpConvertHelper(TypeDescription originalSchema, TypeDescription targetSchema) {
     if (!targetSchema.getCategory().isPrimitive()) {
-      if (!originalSchema.getFieldNames().containsAll(targetSchema.getFieldNames())) {
-        return false;
-      }
-      boolean result = true;
+      if (targetSchema.getCategory().equals(TypeDescription.Category.LIST)) {
+        return eligibleForUpConvertHelper(originalSchema.getChildren().get(0), targetSchema.getChildren().get(0));
```
Comment (sv2000) [inline gobblin-compaction/src/main/java/org/apache/gobblin/compaction/mapreduce/orc/OrcUtils.java:495]: Should we do a null check here? What happens if originalSchema.getChildren() returns null?

File: gobblin-compaction/src/test/java/org/apache/gobblin/compaction/mapreduce/orc/OrcUtilsTest.java [hunk 2]
```diff
@@ -294,6 +294,26 @@ public void testOrcStructProjection() throws Exception {
...
...ing,testSegmentId:string,auditstone:struct<time:bigint,server:string,instance:string,appName:string,messageId:binary,auditVersion:int,fabricUrn:string,clusterConnectionString:string>,pageInstance:struct<pageUrn:string,trackingId:binary>,clientApplicationInstance:struct<applicationUrn:string,version:string,trackingId:binary>,originSource:string,sessionUrn:string,traceData:struct<treeId:binary,requestId:int,taskId:int,rpcTrace:string,forceTraceEnabled:boolean,context:map<string,string>,scaleFactor:double>>,requeststone:struct<browserId:string,sessionId:string,ip:string,pageKey:string,path:string,locale:string,interfaceLocale:string,trackingCode:string,referer:string,userAgent:string,ipAsBytes:binary,requestProtocol:string,requestDomain:string>,statusCode:int,bodyDataAnnotationGroups:array<struct<annotations:array<struct<domainType:struct<entityId:int,attributeId:int>,sourceUrn:string>>,groupId:int,childrenGroupIds:array<int>,hashedSchemaName:int>>,resourceName:string,info:struct<rm:string,actionOrFinderName:string,schemaName:string>,xId:struct<tenantUrn:string,eUrn:string,memberUrn:string,cloudUrn:string,clientId:string,userMetadata:struct<emad:string,accountAgeInSeconds:bigint>>>");
```
Comment (sv2000) [inline gobblin-compaction/src/test/java/org/apache/gobblin/compaction/mapreduce/orc/OrcUtilsTest.java:312]: Two points: 1. Maybe avoid exposing LinkedIn internal schemas outside, and 2. From the schema definitions, it is hard to see what exactly is being tested. Can you add a comment explaining the difference between struct_a and struct_b and why eligibileForUpConvert should return true?

File: gobblin-compaction/src/main/java/org/apache/gobblin/compaction/mapreduce/orc/OrcUtils.java (hunk 1 above)
Comment (autumnust) [reply gobblin-compaction/src/main/java/org/apache/gobblin/compaction/mapreduce/orc/OrcUtils.java:495]: Will add a safety-guard.

File: gobblin-compaction/src/test/java/org/apache/gobblin/compaction/mapreduce/orc/OrcUtilsTest.java [hunk 3]
```diff
@@ -294,6 +294,21 @@ public void testOrcStructProjection() throws Exception {
     Assert.assertEquals(projectColumnStruct, projectedStructExpectedValue);
   }
 
+  @Test
+  public void complextTypeEligibilityCheck() throws Exception {
```
Comment (sv2000) [inline gobblin-compaction/src/test/java/org/apache/gobblin/compaction/mapreduce/orc/OrcUtilsTest.java:298]: minor typo: complextType -> complexType.

File: gobblin-compaction/src/main/java/org/apache/gobblin/compaction/mapreduce/orc/OrcUtils.java [hunk 4]
```diff
@@ -491,18 +490,38 @@ public static WritableComparable createValueRecursively(TypeDescription schema)
...
-      if (!originalSchema.getFieldNames().containsAll(targetSchema.getFieldNames())) {
-        return false;
-      }
-      boolean result = true;
+      if (targetSchema.getCategory().equals(TypeDescription.Category.LIST)) {
+        Preconditions
+            .checkArgument(originalSchema.getChildren() != null, "Illegal format of ORC schema as:" + targetSchema);
+        return eligibleForUpConvertHelper(originalSchema.getChildren().get(0), targetSchema.getChildren().get(0));
+      } else if (targetSchema.getCategory().equals(TypeDescription.Category.MAP)) {
+        Preconditions
```
Comment (sv2000) [inline gobblin-compaction/src/main/java/org/apache/gobblin/compaction/mapreduce/orc/OrcUtils.java:498]: Should the error message be: "Illegal format of ORC schema as:" + originalSchema?

**Pull request closed without merging**